File: BaihuTest/page/albuminfo/ModelInfoViewController.m
//
//  AlbumInfoViewController.m
//  BaihuTest
//
//  Created by liudong on 2020/6/8.
//  Copyright © 2020 liudong. All rights reserved.
//

#import "ModelInfoViewController.h"
#import <Masonry.h>
#import <SDWebImage.h>
#import "AppConfig.h"
#import "DimenAdapter.h"

@interface ModelInfoViewController ()
@property (nonatomic, strong, readwrite) Model *model;
@property (nonatomic, strong, readwrite) UIImageView *topImage;
@property(nonatomic,strong,readwrite)UIView* topBar;
@property(nonatomic,strong,readwrite)UIButton* backIcon;
@end

@implementation ModelInfoViewController

+ (instancetype)initWithModel:(Model *)model {
    ModelInfoViewController *vc = [[ModelInfoViewController alloc]init];
    vc.model = model;
    return vc;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    _topBar = [[UIView alloc]initWithFrame: CGRectZero];
    _topBar.backgroundColor = [UIColor whiteColor];
    [self.view addSubview:_topBar];
    [_topBar mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(self.view);
        make.top.mas_equalTo(self.view.mas_top).offset(STATUSBAR_HEIGHT);
        make.height.mas_equalTo(NAVIGATIONBAR_HEIGHT);
    }];
    _backIcon = [[UIButton alloc]init];
    [_backIcon setImage:[UIImage imageNamed:@"back"] forState:UIControlStateNormal];
    [_backIcon addTarget:self action:@selector(popPage) forControlEvents:UIControlEventTouchUpInside];
    [_topBar addSubview:_backIcon];
    [_backIcon mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(_topBar.mas_centerY);
        make.left.equalTo(_topBar.mas_left).offset(UI(15));
    }];
    
    
    
    [self.view addSubview:self.topImage];
    [self.topImage mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(self.view);
        make.top.mas_equalTo(_topBar.mas_bottom);
        make.height.mas_equalTo(UI(200));
    }];
    NSString *converUrl =   [[AppConfig getInstance]getPhotoWholeUrl:_model.cover_image_url isThumb:NO];
    [self.topImage sd_setImageWithURL:[NSURL URLWithString:converUrl]];
}

- (UIImageView *)topImage {
    if (!_topImage) {
        _topImage = [[UIImageView alloc]init];
    }
    return _topImage;
}

@end

File: BaihuTest/page/albuminfo/ModelInfoViewController.swift
//
//  ModelInfoViewController.swift
//  BaihuTest
//
//  Created by liudong on 2020/6/30.
//  Copyright © 2020 liudong. All rights reserved.
//

import Alamofire
import UIKit

class ModelInfoViewController: DetailInfoControllerBridgeViewController, AlbumInfoForModelOrTagProtocol {
    func paramsForCount() -> [AnyHashable: Any] {
        return ["model_id": (model as! Model).id]
    }

    func paramsForAlbumList() -> [AnyHashable: Any] {
        return ["model_id": (model as! Model).id!, "page": page, "size": pageSize]
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        super.protocolDelegate = self
        //        TODO: 添加模特顶部view
        let nameBg = UIView()
        nameBg.backgroundColor = UIColor(hexString: "#66000000")
        nameBg.layer.cornerRadius = 25
        nameBg.layer.masksToBounds = true
        let nameLabel = UILabel()
        nameLabel.text = (model as! Model).name
        nameLabel.font = UIFont(name: "PingFang-SC-Bold", size: 18)

        nameLabel.textColor = UIColor.white
        nameBg.addSubview(nameLabel)
        nameLabel.snp.makeConstraints { maker in
            maker.center.equalToSuperview()
        }
        view.addSubview(nameBg)
        nameBg.snp.makeConstraints { maker in
            maker.width.equalTo(125)
            maker.height.equalTo(50)
            maker.left.equalToSuperview().offset(60)
            maker.top.equalTo(backIcon.snp.bottom).offset(23)
        }
        view.addSubview(avatarIcon)
        avatarIcon.snp.makeConstraints { maker in
            maker.size.equalTo(70)
            maker.left.equalToSuperview().offset(15)
            maker.top.equalTo(backIcon.snp.bottom).offset(13)
        }
        initTags((model as! Model).tags)
        view.addSubview(descriptionView)
        descriptionView.snp.makeConstraints { maker in
            maker.top.equalTo(topImage.snp.bottom).offset(20)
            maker.left.equalToSuperview().offset(15)
            maker.right.equalToSuperview().offset(-15)
        }

        fetchPhotosCount()
        view.addSubview(pictCount)
        pictCount.snp.makeConstraints { make in
            make.top.equalTo(descriptionView.snp.bottom).offset(24)
            make.left.equalToSuperview().offset(17)
        }

        view.addSubview(photoList)
        photoList.snp.makeConstraints { maker in
            maker.top.equalTo(pictCount.snp.bottom).offset(15)
            maker.left.right.equalToSuperview()
            maker.bottom.equalToSuperview()
        }
        fetchPhotos(page: page, pageSize: pageSize)
    }

    func albumsCountUrl() -> String {
        return UrlConstants.getAlbumCountForModel()
    }

    func albumListUrl() -> String {
        return UrlConstants.getAlbumsForModel()
    }

    lazy var avatarIcon: UIImageView = {
        let imageview = UIImageView()
        imageview.sd_setImage(with: (model as! Model).avatar_image_url.keyStrToUrl(isThumb: true), completed: nil)
        imageview.layer.cornerRadius = 35
        imageview.layer.masksToBounds = true
        return imageview
    }()

    lazy var descriptionView: UILabel = {
        let label = UILabel()
        label.numberOfLines = 0
        label.text = (model as! Model).summary
        label.sizeToFit()
        label.textColor = UIColor(hexString: "#FF999999")
        label.font = UIFont(name: "PingFang-SC-Medium", size: 14)
        return label
    }()

    func initTags(_ tagStr: String) {
        var tags = [String]()

        tagStr.components(separatedBy: "|").forEach { it in
            tags.append(it)
        }

        let tagView = CustomTagView(tags: tags, frame: CGRect.zero)
        view.addSubview(tagView)
        tagView.snp.makeConstraints { maker in
            maker.left.equalToSuperview().offset(20)
            maker.right.equalToSuperview().offset(-20)
            maker.height.equalTo(100)
            maker.top.equalTo(avatarIcon.snp.bottom).offset(20)
        }
        print("")
    }
}
